### PR TITLE
fix: using configuration for tls and ca secret names

### LINKED ISSUE #521: Resource name mismatch in the code vs charts
# Bug description

The CA secret name in the code has been hard-coded as `capsule-ca` (ref [here](https://github.com/clastix/capsule/blob/7fefe4f6dedcdf5c600f4edc0f55657ab4bfbb5b/controllers/secret/const.go#L10)), which is what is used to search for the _secret_ that provides the CA for the controller manager.

But in the chart, the name of the secret is templated as `name: {{ include "capsule.secretCaName" . }}` (ref [here](https://github.com/clastix/capsule/blob/7fefe4f6dedcdf5c600f4edc0f55657ab4bfbb5b/charts/capsule/templates/ca.yaml#L10)). This is included from the __helpers.tpl_ as `{{- printf "%s-ca" (include "capsule.fullname" .) -}}` (ref [here](https://github.com/clastix/capsule/blob/7fefe4f6dedcdf5c600f4edc0f55657ab4bfbb5b/charts/capsule/templates/_helpers.tpl#L127)) which is an ovveride with the release name.

So if I release my chart with `releaseName=multi-tenancy-manager` then the secret name will be rendered as multi-tenancy-manager-capsule-ca`. This is not the secret name the controller is looking for in [here](https://github.com/clastix/capsule/blob/1d10bcab1ea75b85585db3b09ddae718e7c85f42/main.go#L156).

# How to reproduce

helm template multi-tenancy-manager clastix/capsule --include-crds

# Logs

```
{"level":"error","ts":"2022-03-01T23:40:56.690Z","logger":"setup","msg":"unable to get Capsule CA secret","error":"secrets \"capsule-ca\" not found","stacktrace":"main.main\n\t/workspace/main.go:158\nruntime.main\n\t/usr/local/go/src/runtime/proc.go:225"}
```

# Additional context

- Capsule version: `v0.1.1 1bbaebb`
- Helm Chart version: `0.1.6`
- Kubernetes version: `v1.21.5`

## I can submit a PR on the charts repo by hard-coding the ca secret name to capsule-ca. Please let me know your suggestion.

### PR DESCRIPTION
**Pull request by davideimola**
fixes #521

### CODE REVIEW

File: pkg/configuration/configuration.go [hunk 1]
```diff
@@ -12,7 +12,13 @@ import (
...
+	CASecretName() string
+	TLSSecretName() string
 	UserGroups() []string
 	ForbiddenUserNodeLabels() *capsulev1beta1.ForbiddenListSpec
 	ForbiddenUserNodeAnnotations() *capsulev1beta1.ForbiddenListSpec
 }
+
+const CASecretName = "capsule-ca"
+
+const TLSSecretName = "capsule-tls"
```
Comment (prometherion) [inline pkg/configuration/configuration.go:24]: Let's move the constants after the imports, collapsing into a single declaration.

```suggestion
const (
    CASecretName  = "capsule-ca"
    TLSSecretName = "capsule-tls"
)
```

**Review by prometherion — CHANGES_REQUESTED** [review]

**Review by gkarthiks — APPROVED** [review]
Apart from the cosmetic comment from @prometherion, this looks great to me. Looking forward for a version of capsule with this change.

**Comment by davideimola** [comment]
@prometherion let me know something about the documentation

**Review by prometherion — APPROVED** [review]

**Comment by prometherion** [comment]
Tests are flaky, not a big deal.

I'll test locally ASAP, overall LGTM.

**Comment by davideimola** [comment]
Thanks @prometherion! Let me know if you need anything!!

**Comment by maxgio92** [comment]
Hello @davideimola , I tested with this configuration, installed Capsule through Helm chart (I patched it to include the new annotations on the `CapsuleConfiguration`).

The failure on the search of the CA `Secret` seems solved from my test:
```
pippo-capsule-controller-manager-56b5956dc6-lj46d manager {"level":"info","ts":"2022-03-14T11:28:32.245Z","logger":"controllers.CA","msg":"Reconciling CA Secret","Request.Namespace":"capsule-system","Request.Name":"pippo-capsule-ca"}

pippo-capsule-controller-manager-56b5956dc6-lj46d manager {"level":"info","ts":"2022-03-14T11:28:36.963Z","logger":"controllers.CA","msg":"Handling CA Secret","Request.Namespace":"capsule-system","Request.Name":"pippo-capsule-ca"}
pippo-capsule-controller-manager-56b5956dc6-lj46d manager {"level":"info","ts":"2022-03-14T11:28:36.963Z","logger":"controllers.CA","msg":"Updating CA secret with new PEM and RSA","Request.Namespace":"capsule-system","Request.Name":"pippo-capsule-ca"}
```

Instead, the search of the certificate and key `Secret` seems not to succeed, for the waiting `Job`:
```
pippo-capsule-waiting-certs-frvl8 post-install-job Error from server (NotFound): secrets "capsule-tls" not found
pippo-capsule-waiting-certs-frvl8 post-install-job waiting Capsule to be up and running...
```

To reproduce the test, the install command:
```
helm install pippo-capsule ./charts/capsule -n capsule-system --create-namespace -f values-test.yaml
```

The new `CapsuleConfiguration` template:
```
apiVersion: capsule.clastix.io/v1alpha1
kind: CapsuleConfiguration
metadata:
  name: default
  labels:
  {{- include "capsule.labels" . | nindent 4 }}
  annotations:
    capsule.clastix.io/ca-secret-name: {{ include "capsule.secretCaName" . }}
    capsule.clastix.io/tls-secret-name: {{ include "capsule.secretTlsName" . }}
  {{- with .Values.customAnnotations }}
    {{- toYaml . | nindent 4 }}
  {{- end }}
spec:
  forceTenantPrefix: {{ .Values.manager.options.forceTenantPrefix }}
  userGroups:
{{- range .Values.manager.options.capsuleUserGroups }}
    - {{ . }}
{{- end}}
  protectedNamespaceRegex: {{ .Values.manager.options.protectedNamespaceRegex | quote }}
```
and the templated `CapsuleConfiguration` result:
```
apiVersion: capsule.clastix.io/v1alpha1
kind: CapsuleConfiguration
metadata:
  annotations:
    capsule.clastix.io/ca-secret-name: pippo-capsule-ca
    capsule.clastix.io/tls-secret-name: pippo-capsule-tls
    meta.helm.sh/release-name: pippo-capsule
    meta.helm.sh/release-namespace: capsule-system
  ...
  labels:
    ...
  name: default
 ...
  spec:
    forceTenantPrefix: false
    protectedNamespaceRegex: ""
    userGroups:
    - capsule.clastix.io
```

Moreover, after the CA retrieval, the `MutatingWebhookConfiguration` and `ValidatingWebhookConfiguration` retrieval seem to fail:

```
pippo-capsule-controller-manager-56b5956dc6-lj46d manager {"level":"info","ts":"2022-03-14T11:28:32.245Z","logger":"controllers.CA","msg":"Reconciling CA Secret","Request.Namespace":"capsule-system","Request.Name":"pippo-capsule-ca"}

pippo-capsule-controller-manager-56b5956dc6-lj46d manager {"level":"info","ts":"2022-03-14T11:28:36.963Z","logger":"controllers.CA","msg":"Handling CA Secret","Request.Namespace":"capsule-system","Request.Name":"pippo-capsule-ca"}
pippo-capsule-controller-manager-56b5956dc6-lj46d manager {"level":"info","ts":"2022-03-14T11:28:36.963Z","logger":"controllers.CA","msg":"Updating CA secret with new PEM and RSA","Request.Namespace":"capsule-system","Request.Name":"pippo-capsule-ca"}
pippo-capsule-controller-manager-56b5956dc6-lj46d manager {"level":"error","ts":"2022-03-14T11:28:36.975Z","logger":"controllers.CA","msg":"cannot retrieve ValidatingWebhookConfiguration","Request.Namespace":"capsule-system","Request.Name":"pippo-capsule-ca","error":"ValidatingWebhookConfiguration.admissionregistration.k8s.io \"capsule-validating-webhook-configuration\" not found","stacktrace":"k8s.io/client-go/util/retry.OnError.func1\n\t/go/pkg/mod/k8s.io/client-go@v0.22.0/util/retry/util.go:51\nk8s.io/apimachinery/pkg/util/wait.ConditionFunc.WithContext.func1\n\t/go/pkg/mod/k8s.io/apimachinery@v0.22.0/pkg/util/wait/wait.go:217\nk8s.io/apimachinery/pkg/util/wait.runConditionWithCrashProtectionWithContext\n\t/go/pkg/mod/k8s.io/apimachinery@v0.22.0/pkg/util/wait/wait.go:230\nk8s.io/apimachinery/pkg/util/wait.runConditionWithCrashProtection\n\t/go/pkg/mod/k8s.io/apimachinery@v0.22.0/pkg/util/wait/wait.go:223\nk8s.io/apimachinery/pkg/util/wait.ExponentialBackoff\n\t/go/pkg/mod/k8s.io/apimachinery@v0.22.0/pkg/util/wait/wait.go:418\nk8s.io/client-go/util/retry.OnError\n\t/go/pkg/mod/k8s.io/client-go@v0.22.0/util/retry/util.go:50\nk8s.io/client-go/util/retry.RetryOnConflict\n\t/go/pkg/mod/k8s.io/client-go@v0.22.0/util/retry/util.go:104\ngithub.com/clastix/capsule/controllers/secret.CAReconciler.UpdateValidatingWebhookConfiguration\n\t/workspace/controllers/secret/ca.go:81\ngithub.com/clastix/capsule/controllers/secret.CAReconciler.Reconcile.func2\n\t/workspace/controllers/secret/ca.go:171\ngolang.org/x/sync/errgroup.(*Group).Go.func1\n\t/go/pkg/mod/golang.org/x/sync@v0.0.0-20210220032951-036812b2e83c/errgroup/errgroup.go:57"}
pippo-capsule-controller-manager-56b5956dc6-lj46d manager {"level":"error","ts":"2022-03-14T11:28:36.975Z","logger":"controllers.CA","msg":"cannot retrieve MutatingWebhookConfiguration","Request.Namespace":"capsule-system","Request.Name":"pippo-capsule-ca","error":"MutatingWebhookConfiguration.admissionregistration.k8s.io \"capsule-mutating-webhook-configuration\" not found","stacktrace":"k8s.io/client-go/util/retry.OnError.func1\n\t/go/pkg/mod/k8s.io/client-go@v0.22.0/util/retry/util.go:51\nk8s.io/apimachinery/pkg/util/wait.ConditionFunc.WithContext.func1\n\t/go/pkg/mod/k8s.io/apimachinery@v0.22.0/pkg/util/wait/wait.go:217\nk8s.io/apimachinery/pkg/util/wait.runConditionWithCrashProtectionWithContext\n\t/go/pkg/mod/k8s.io/apimachinery@v0.22.0/pkg/util/wait/wait.go:230\nk8s.io/apimachinery/pkg/util/wait.runConditionWithCrashProtection\n\t/go/pkg/mod/k8s.io/apimachinery@v0.22.0/pkg/util/wait/wait.go:223\nk8s.io/apimachinery/pkg/util/wait.ExponentialBackoff\n\t/go/pkg/mod/k8s.io/apimachinery@v0.22.0/pkg/util/wait/wait.go:418\nk8s.io/client-go/util/retry.OnError\n\t/go/pkg/mod/k8s.io/client-go@v0.22.0/util/retry/util.go:50\nk8s.io/client-go/util/retry.RetryOnConflict\n\t/go/pkg/mod/k8s.io/client-go@v0.22.0/util/retry/util.go:104\ngithub.com/clastix/capsule/controllers/secret.CAReconciler.UpdateMutatingWebhookConfiguration\n\t/workspace/controllers/secret/ca.go:100\ngithub.com/clastix/capsule/controllers/secret.CAReconciler.Reconcile.func1\n\t/workspace/controllers/secret/ca.go:168\ngolang.org/x/sync/errgroup.(*Group).Go.func1\n\t/go/pkg/mod/golang.org/x/sync@v0.0.0-20210220032951-036812b2e83c/errgroup/errgroup.go:57"}
pippo-capsule-controller-manager-56b5956dc6-lj46d manager {"level":"error","ts":"2022-03-14T11:28:37.086Z","logger":"controller-runtime.manager.controller.secret","msg":"Reconciler error","reconciler group":"","reconciler kind":"Secret","name":"pippo-capsule-ca","namespace":"capsule-system","error":"ValidatingWebhookConfiguration.admissionregistration.k8s.io \"capsule-validating-webhook-configuration\" not found","stacktrace":"sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).processNextWorkItem\n\t/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.9.5/pkg/internal/controller/controller.go:253\nsigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Start.func2.2\n\t/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.9.5/pkg/internal/controller/controller.go:214"}
```

But maybe this last one can be solved at Helm chart level. Do we need to name the webhook configuration with custom values?

**Comment by davideimola** [comment]
> Instead, the search of the certificate and key `Secret` seems not to succeed, for the waiting `Job`:


I'll try to figure it out why this things brokes. I did not find any fixed name inside the code, so it is a bit strange...

**Comment by davideimola** [comment]
Hi folks,
I'm wondering if the TLS fails discovered by @maxgio92 may seem related to that part

https://github.com/clastix/capsule/blob/7fefe4f6dedcdf5c600f4edc0f55657ab4bfbb5b/charts/capsule/templates/post-install-job.yaml#L1

@prometherion @maxgio92 do you have any other suggestions?

**Comment by prometherion** [comment]
> Hi folks, I'm wondering if the TLS fails discovered by @maxgio92 may seem related to that part
> 
> https://github.com/clastix/capsule/blob/7fefe4f6dedcdf5c600f4edc0f55657ab4bfbb5b/charts/capsule/templates/post-install-job.yaml#L1

Yes, that's the part we have to fix: we have to update Charts in order to reflect all the changes.

**Comment by prometherion** [comment]
If you agree @davideimola let's provide an option to specify a different Mutating/Validating Webhook Configuration as we're doing with Secret, and at the same time, fixing the Charts in order to wait for the certificates.

**Comment by davideimola** [comment]
> > Hi folks, I'm wondering if the TLS fails discovered by @maxgio92 may seem related to that part
> > https://github.com/clastix/capsule/blob/7fefe4f6dedcdf5c600f4edc0f55657ab4bfbb5b/charts/capsule/templates/post-install-job.yaml#L1
> 
> Yes, that's the part we have to fix: we have to update Charts in order to reflect all the changes.

Ok, what do you suggest to do in order to fix that line? 

> If you agree @davideimola let's provide an option to specify a different Mutating/Validating Webhook Configuration as we're doing with Secret, and at the same time, fixing the Charts in order to wait for the certificates.

Yes, I can absolutely make it a try.

**Comment by prometherion** [comment]
@davideimola I guess you can easily fix your issue with Helm charts with the following patch.

```patch
diff --git a/charts/capsule/templates/post-install-job.yaml b/charts/capsule/templates/post-install-job.yaml
index 3a40bec..58358aa 100644
--- a/charts/capsule/templates/post-install-job.yaml
+++ b/charts/capsule/templates/post-install-job.yaml
@@ -1,4 +1,4 @@
-{{- $cmd := "while [ -z $$(kubectl -n $NAMESPACE get secret capsule-tls -o jsonpath='{.data.tls\\\\.crt}') ];" -}}
+{{- $cmd := printf "while [ -z $$(kubectl -n $NAMESPACE get secret %s -o jsonpath='{.data.tls\\\\.crt}') ];" (include "capsule.secretCaName" .) -}}
 {{- $cmd = printf "%s do echo 'waiting Capsule to be up and running...' && sleep 5;" $cmd -}}
 {{- $cmd = printf "%s done" $cmd -}}
 apiVersion: batch/v1
```

Helm compiles, let's see if it works in a real environment :)

**Comment by davideimola** [comment]
Thanks @prometherion! The CI fails again 😅
Let me know if it might be something related to my changes!

**Review by prometherion — APPROVED** [review]

**Pull request merged**